### ISSUE TITLE
Further glitches in the list of known languages/locales

### ISSUE THREAD

**Issue by utelle** [wxWidgets/wxWidgets#23437]
In the course of streamlining the process to regenerate the file `misc/languages/langtabl.txt` containing the list of languages/locales with associated `wxLANGUAGE_*` symbols, I discovered 2 glitches:

1. The symbol `wxLANGUAGE_CENTRAL_KURDISH` is listed 2 times, for `ku` and `ku-Arab`.
2. The symbol `wxLANGUAGE_JAVANESE`  is listed 2 times, for `jv` and `jv-Latn`.

These glitches were present since the list of languages was extended (from 234 to 877 entries) in wxWidgets version 3.1.6. This duplication went undetected, because the script `genlang.py` silently eliminates double entries of language symbols.

The affected locales are rather "exotic" and I doubt that any productive wxWidgets application is/was affected. Therefore I would not backport the necessary fixes to version 3.2, because fixes would inevitably introduce new symbols.

Additionally, I'd like to bring to your attention that the list of languages associates 3 generic language entries with certain region specific entries where the list of known locales under Windows uses different associations:

1. Chinese `zh` is associated with `zh_TW` (Taiwan), while under Windows it is associated with `zh_CN` (China).
2. English `en` is associated with `en_GB` (Great Britain, United Kingdom), while under Windows it is associated with `en_US` (USA).
3. Portuguese `pt` is associated with `pt_PT` (Portugal), while under Windows it is associated with `pt_BR` (Brazil).

For wx the associations (`zh` <=> `zh_TW`), (`en` <=> `en_GB`), and (`pt` <=> `pt_PT`) were chosen for compatibilty reasons with prior wxWidgets versions. However, most operating systems use the associations  (`zh` <=> `zh_CN`), (`en` <=> `en_US`), and (`pt` <=> `pt_BR`).

The big question is whether we should keep the current wx associations (for compatibility) or whether we should replace them by the more common ones (as used in most operating systems). Opinions?

**Comment by vadz** [comment]
I am not sure at all what's the right thing to do here, so I'd rather keep the definitions as they are and just document somewhere that `wxLANGUAGE_ENGLISH` etc might not have the expected definition and that the more precise versions such as `wxLANGUAGE_ENGLISH_US` should be used instead.

**Comment by utelle** [comment]
> I am not sure at all what's the right thing to do here, so I'd rather keep the definitions as they are

Keep "as they are" where? As currently in the wxWidgets list? Or as in the Windows defaults?

The problem I see with keeping the current definition in the wxWidgets language list is that one could get inconsistent results. If `wxUILocale` is used directly asking for locale `en` would create a locale corresponding to `en-US` while using `wxLocale` would map `en` to `en-GB`. Similar for the other mentioned ids. 

The downside of changing the definitions in wxWidgets is that existing applications could experience differences when upgrading to a new wxWidgets version.

OTOH new applications should prefer to use locale ids like `en`, `en-US` etc instead of `wxLANGUAGE` symbols.

> and just document somewhere that `wxLANGUAGE_ENGLISH` etc might not have the expected definition and that the more precise versions such as `wxLANGUAGE_ENGLISH_US` should be used instead.

Most likely the documentation should be improved in that respect.

**Comment by utelle** [comment]
> > Keep "as they are" where? As currently in the wxWidgets list?
> 
> Yes.

Ok.

> Windows mapping probably makes more sense (although I'm not really convinced about Portuguese), but I have a feeling that we shouldn't even try to determine what is the "primary" variant of any language.

I agree. That would be a fruitless discussion. However, the problem is that one may get inconsistent behaviour, at least under Windows. For the language `wxLANGUAGE_ENGLISH` the locale id `en` would be used internally in `wxUILocale` to set the locale, and that would actually correspond to `en-US` (while the wxWidgets language database associates the symbol with `en-GB`).

Under Linux one would not have this problem, because the `wxUILocale` implementation uses internally method `GetCanonicalWithRegion` to determine the locale id.

I don't know what would happen under Mac OS X.

> > The problem I see with keeping the current definition in the wxWidgets language list is that one could get inconsistent results. If `wxUILocale` is used directly asking for locale `en` would create a locale corresponding to `en-US` while using `wxLocale` would map `en` to `en-GB`. Similar for the other mentioned ids.
> 
> This is definitely unexpected, but I think it's just another reason to tell people to use `en_US` explicitly in the first place if this is what they want.

If we want consistent behaviour it would be necessary to change the `wxUILocaleImpl` implementation for Windows to use the _canonical name with region_ when a generic language symbol like `wxLANGUAGE_ENGLISH` is used to create the locale.

> Is there any major difference between using `en-US` and `wxLANGUAGE_ENGLISH_US`? I admit I'm still partial to the latter just because any typos in it would be detected at compilation-, rather than run-, time.

No, `en-US` is equivalent to `wxLANGUAGE_ENGLISH_US`.

**Comment by BrianNixon** [comment]
> 2. The symbol `wxLANGUAGE_JAVANESE` is listed 2 times, for `jv` and `jv-Latn`.

Is that why the `CanonicalName` and `CanonicalRef` are the wrong way round? (Specific `jv-Latn` points to generic `jv` rather than the other way round. Not sure if there are any other instances of this.)

Also, on the general theme: is it the intention to keep `langtabl.txt`/​`languageinfo.cpp` encoded entirely in US-⁠ASCII to avoid source-code charset issues in different compilers? If so, there are some entries slipping through in UTF-⁠8 (`"Serbian (Cyrillic)","српски (Србија)"` is one example I noticed; I haven’t checked exhaustively for others.)

**Comment by utelle** [comment]
> > 2. The symbol `wxLANGUAGE_JAVANESE` is listed 2 times, for `jv` and `jv-Latn`.
> 
> Is that why the `CanonicalName` and `CanonicalRef` are the wrong way round? (Specific `jv-Latn` points to generic `jv` rather than the other way round. Not sure if there are any other instances of this.)

No. `CanonicalRef` usually is only set for locales without a specified region. If possible, it points to an entry with the main region where the language is spoken.

That `jv-Latn` points to `jv` is wrong (but does no harm). However, thanks for mentioning this issue. I will fix that, too. Actually, this glitch is a remain of a manual modification.

AFAICT no other entry is affected.

> Also, on the general theme: is it the intention to keep `langtabl.txt`/​`languageinfo.cpp` encoded entirely in US-⁠ASCII to avoid source-code charset issues in different compilers?

Yes. Not all compilers assume UTF-8 encoding for the sources. Therefore non-ASCII characters are converted to an octal representation.

> If so, there are some entries slipping through in UTF-⁠8 (`"Serbian (Cyrillic)","српски (Србија)"` is one example I noticed; I haven’t checked exhaustively for others.)

I had to manually add a small number of entries for compatibility with prior wxWidgets versions, among them entries for `sr-Cyrl-YU` (_Serbian_ in _Yugoslavia_ - note that Yugoslavia doesn't exist anymore, that is, this entry exists only for compatibility reasons) and `ku-TR` (_Kurdish_ in _Turkiye_). Unfortunately, I applied the conversion to the wrong data fields. This will be fixed, of course.

I checked all manual entries, and the above 2 entries were the only ones without correct conversion. All other entries are converted automatically.

I will make a PR as soon as I have finished the streamlining of the regeneration process.
